### PR TITLE
Bah 790 - API related changes.

### PR DESCRIPTION
**Pull request by buvaneswari-arun**

### CODE REVIEW

File: api/src/main/java/org/bahmni/module/bahmni/ie/apps/dao/impl/BahmniFormDaoImpl.java [hunk 1]
```diff
@@ -75,13 +77,29 @@ public BahmniFormDaoImpl(SessionFactory sessionFactory) {
...
-				"F.name as name, F.uuid as uuid, F.version as version, F.published as published " +
+				"F.name as name, F.formId as id, F.uuid as uuid, F.version as version, F.published as published " +
 				"from FormResource FR right outer join FR.form F with FR.datatypeClassname!=" +
 				"'org.bahmni.customdatatype.datatype.FileSystemStorageDatatype' %s %s %s %s " +
 				"order by F.name asc, F.version desc", whereClause, publishedQueryCondition, retiredQueryCondition, formNameQueryCondition))
 				.setResultTransformer( Transformers.aliasToBean(BahmniForm.class));
 
-		return query.list();
+		List<BahmniForm> result =  query.list();
+		List<BahmniForm> finalList = new ArrayList<BahmniForm>();
```
Comment (binduak) [inline api/src/main/java/org/bahmni/module/bahmni/ie/apps/dao/impl/BahmniFormDaoImpl.java]: @buvaneswari-arun can _finalList_ be changed to more relevant name ?

File: api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/BahmniFormService.java [hunk 2]
```diff
@@ -20,4 +21,8 @@
     ExportResponse formDetailsFor(List<String> formUuids);
 
     BahmniFormResource saveFormNameTranslation(BahmniFormResource bahmniFormResource, String referenceFormUuid);
+
+    Form getFormDetailsFromFormName(String formName, String formVersion);
```
Comment (binduak) [inline api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/BahmniFormService.java:25]: @buvaneswari-arun can _getFormDetailsFromFormName_ changed to _getFormDetails**For**FormName_ ?

File: api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    @Override
+    @Transactional
+    public List<FormPrivilege> saveFormPrivileges(List<FormPrivilege> formPrivileges) {
+        log.info("Inside BahmniFormPrivilegesServiceImpl -->saveFormPrivileges " + formPrivileges);
+        List<FormPrivilege> resultList = new ArrayList<FormPrivilege>();
+        List<FormPrivilege> privilegesList = new ArrayList<FormPrivilege>();
+        List<FormPrivilege> oldPrivilegeList = new ArrayList<FormPrivilege>();
+
+//        if (formPrivileges.size() == 1 && formPrivileges.get(0).getPrivilegeName().equalsIgnoreCase("")) {
+//            oldPrivilegeList = getAllPrivilegesForForm(formPrivileges.get(1).getFormId());
```
Comment (binduak) [inline api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java]: @buvaneswari-arun is this commented code is required to be here ?

File: api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+//                    resultList.add(bahmniFormPrivilegeDao.deleteFormPrivilege(oldPrivilegeList.get(i)));
+//                }
+//            }
+//            return resultList;
+//        } else {
+
+            Iterator privilegeItr = formPrivileges.iterator();
+            int formId = 0;
+            String formVersion = "";
+            log.info("Inside BahmniFormPrivilegesServiceImpl -->saveFormPrivileges " + privilegeItr);
```
Comment (binduak) [inline api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java]: @buvaneswari-arun are these logs added intentionally ?

File: api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+                        tempPrivilege.setEditable((Boolean) temp.get("editable"));
+                    } else if (tempEntrySet.getKey().toString().equalsIgnoreCase("privilegeName")) {
+                         String privilegeName = temp.get("privilegeName").toString();
+                         tempPrivilege.setPrivilegeName(privilegeName);
+                    } else if (tempEntrySet.getKey().toString().equalsIgnoreCase("viewable")) {
+                        tempPrivilege.setViewable((Boolean) temp.get("viewable"));
+                    }else if (tempEntrySet.getKey().toString().equalsIgnoreCase("formVersion")) {
+                        tempPrivilege.setFormVersion((String) temp.get("formVersion"));
+                        formVersion = tempPrivilege.getFormVersion();
+                    }
```
Comment (binduak) [inline api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java:84]: @buvaneswari-arun can this method split into multiple private methods. It makes the method more readable

File: api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+                    if (privilegesList != null) {
+                        for(int i=0;i<privilegesList.size();i++){
+                            resultList.add(saveFormPrivilege(privilegesList.get(i)));
+                        }
+                    }
+            }
+        return resultList;
+
+    }
+        @Transactional(readOnly = true)
```
Comment (binduak) [inline api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java:106]: @buvaneswari-arun can you please check the code formatting here

File: api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    public FormPrivilege getFormPrivilege(String privilgeName , Integer formId) throws APIException {
+       return bahmniFormPrivilegeDao.getFormPrivilege(privilgeName,formId);
+   }
+    public void deleteAllThePrivilegesFromDB( Integer formId , String formVersion){
```
Comment (binduak) [inline api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java:110]: any specific why the _deleteAllThePrivilegesFromDB_ is public method ?

File: api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormServiceImpl.java [hunk 8]
```diff
@@ -192,6 +192,11 @@ public BahmniFormResource saveFormNameTranslation(BahmniFormResource bahmniFormR
         return new BahmniFormMapper().map(formResource);
     }
 
+    @Override
+    public Form getFormDetailsFromFormName(String formName, String formVersion){
```
Comment (binduak) [inline api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormServiceImpl.java:196]: @buvaneswari-arun same comment as mentioned above

File: api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/BahmniFormService.java (hunk 2 above)
Comment (buvaneswari-arun) [reply api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/BahmniFormService.java:25]: Will take this up as part of next build as it has needs extensive testing. Followup card created.

File: api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java (hunk 3 above)
Comment (buvaneswari-arun) [reply api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java]: Removed

File: api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java (hunk 5 above)
Comment (buvaneswari-arun) [reply api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java:84]: Will take up as part of next build

File: api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java (hunk 6 above)
Comment (buvaneswari-arun) [reply api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java:106]: Done

File: api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java (hunk 7 above)
Comment (buvaneswari-arun) [reply api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java:110]: It is a service method, which may be used for Rest calls, hence preferred to use public.

**Review by binduak — APPROVED** [review]

**Comment by binduak** [comment]
Approved the PR. @buvaneswari-arun can you please share the new JIRA card created to work on the above changes as part of the next build here

File: api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java (hunk 4 above)
Comment (buvaneswari-arun) [reply api/src/main/java/org/bahmni/module/bahmni/ie/apps/service/impl/BahmniFormPrivilegesServiceImpl.java]: Removed

**Comment by buvaneswari-arun** [comment]
BAH-1099  - This card is used to track the test cases and the code review comments that's is pending to be implemented for next code cut in BAH790 functionality.

**Pull request merged**